[ADD][WIP] event_session: New module

chienandalu opened the pull request:
# Create and assign sessions to events

This module allows to create sessions associated with events.

## Usage

You can either:

* Go to Events > Sessions and create some sessions associated with an event.
* Go to an event a use the sessions wizard to create all your sessions according to a schedule.

chienandalu committed:
[ADD] event_session: New module

chienandalu commented:
cc @Tecnativa

sergio-teruel commented on event_session/README.rst:
```diff
@@ -0,0 +1,45 @@
+You can either:
+
+* Go to Events > Sessions and create some sessions associated with an event.
+* Go to an event a use the sessions wizard to create all your sessions according to a schedule.
```
s/a/and

sergio-teruel replied:
s/all your sessions/event

sergio-teruel commented on event_session/README.rst:
```diff
@@ -0,0 +1,45 @@
+======================
+
+* TODO: Seats and registrations management
+* TODO: Improve hours constraints
```
Why this features is in roadmap?

sergio-teruel commented on event_session/__manifest__.py:
```diff
@@ -0,0 +1,21 @@
+# -*- coding: utf-8 -*-
+# Part of Odoo. See LICENSE file for full copyright and licensing details.
```
Fix copyrigth according to https://github.com/OCA/maintainer-tools/blob/master/template/module/__openerp__.py

sergio-teruel commented on event_session/models/event.py:
```diff
@@ -0,0 +1,37 @@
+    session_id = fields.Many2one(
+        comodel_name='event.session',
+        string='Session',
+        ondelete='set null',
```
'set null' is the deafult value so is not necessary,but if you remove an event session that has registrations you will lose the data relative to session. Really do you want that?

chienandalu replied:
Yes, I think it's better to keep the assistant even if the associated session is deleted.

yajo replied:
If that's the case, just remove the line as @sergio-teruel said please

pedrobaeza replied:
This should be `ondelete='restrict'`. You cannot remove a session if you have registration on it.

sergio-teruel commented on event_session/models/event.py:
```diff
@@ -0,0 +1,37 @@
+        string='Sessions',
+    )
+    sessions_count = fields.Integer(
+        compute='_compute_sessions_count'
```
Add string attribute and put a comma at the end.

sergio-teruel commented on event_session/models/event_session.py:
```diff
@@ -0,0 +1,71 @@
+    )
+    start_time = fields.Float(
+        required=True,
+        help="Each session will start at this time (in the event's timezone)",
```
You are in session record not in wizard, so this start time is only one.
s/Each session will/Session start time

sergio-teruel commented on event_session/models/event_session.py:
```diff
@@ -0,0 +1,71 @@
+    )
+    end_time = fields.Float(
+        required=True,
+        help="Each session will end at this time",
```
the same...

sergio-teruel commented on event_session/models/event_session.py:
```diff
@@ -0,0 +1,71 @@
+    @api.multi
+    @api.constrains('start_time', 'end_time')
+    def _avoid_zero_duration(self):
+        for session in self:
```
Keep conventions as:
```python
  # Constraints and onchanges
    @api.constrains('seats_max', 'seats_available')
    def _check_seats_limit(self):
        ...
```
Take a look https://github.com/OCA/maintainer-tools/blob/master/CONTRIBUTING.md

sergio-teruel commented on event_session/models/event_session.py:
```diff
@@ -0,0 +1,71 @@
+    @api.multi
+    @api.constrains('date')
+    def _avoid_out_of_event_date_range(self):
+        for session in self:
```
Same

sergio-teruel commented on event_session/views/event_session_view.xml:
```diff
@@ -0,0 +1,70 @@
+          parent="event.event_main_menu" action="act_event_session_from"
+          groups="event.group_event_user"/>
+
+</odoo>
```
Add blank line

sergio-teruel commented on event_session/views/event_view.xml:
```diff
@@ -0,0 +1,54 @@
+    </field>
+</record>
+
+</odoo>
```
Add blank line

sergio-teruel commented on event_session/wizards/wizard_event_session.py:
```diff
@@ -0,0 +1,211 @@
+        "Session info",
+        required=True,
+        help="It will be generated according to given parameters",
+        default='void',
```
If you want a temporaly name for field you must be use **default='/'**

sergio-teruel commented on event_session/wizards/wizard_event_session.py:
```diff
@@ -0,0 +1,211 @@
+    event_date_begin = fields.Datetime(
+        related="event_id.date_begin",
+        help="Set it up in the event configuration"
+             "Sessions will be generated from this date",
```
This related field must be readonly

sergio-teruel commented on event_session/wizards/wizard_event_session.py:
```diff
@@ -0,0 +1,211 @@
+    event_date_end = fields.Datetime(
+        related="event_id.date_end",
+        help="Set it up in the event configuration"
+             "Sessions will be generated up to this date",
```
This related field must be readonly

sergio-teruel commented on event_session/wizards/wizard_event_session.py:
```diff
@@ -0,0 +1,211 @@
+    event_date_tz = fields.Selection(
+        related="event_id.date_tz",
+        help="Set it up in the event configuration"
+             "Sessions will be generated up to this date",
```
Readonly

sergio-teruel commented on event_session/wizards/wizard_event_session.py:
```diff
@@ -0,0 +1,211 @@
+    def existing_sessions(self, date):
+        """Return existing sessions that match some criteria."""
+        return self.env["event.session"].search(
+            (("event_id", "=", self.event_id.id),
```
domain format is a list that contains tuples as [('a', '=', 1), ()...]

sergio-teruel commented on event_session/wizards/wizard_event_session.py:
```diff
@@ -0,0 +1,211 @@
+            "seats_max": self.event_id.seats_max,
+            "seats_availability": self.event_id.seats_availability,
+        }
+        data.update(values)
```
If you do a method as "_prepare_session_values()" that return a dictionary other modules will can extend this method

sergio-teruel commented on event_session/wizards/wizard_event_session.py:
```diff
@@ -0,0 +1,211 @@
+        weekdays = self.weekdays()
+        if not any(weekdays):
+            raise ValidationError(_("You must select at least one weekday"))
+
```
Remove white blank lines

sergio-teruel commented on event_session/wizards/wizard_event_session.py:
```diff
@@ -0,0 +1,211 @@
+
+        if self.delete_existing_sessions:
+            self.event_id.session_ids.unlink()
+
```
Remove white blank lines

sergio-teruel commented:
don't forget translations and tests :smile:

sergio-teruel commented on event_session/tests/test_session.py:
```diff
@@ -0,0 +1,58 @@
+        # delete previous sessions
+        self.wizard.update({'delete_existing_sessions': True})
+        self.wizard.action_generate_sessions()
+        with self.assertRaises(ValidationError), self.cr.savepoint():
```
Make test for all created methods and all cases..

yajo commented on event_session/wizards/wizard_event_session.py:
```diff
@@ -0,0 +1,251 @@
+    def create_session(self, **values):
+        """Create a new session record with the provided values."""
+        data = {
+            "name": "{} {} - {}".format(
```
You must use a Unicode string for using `.format`

chienandalu replied:
Well, it crashes if I do so whenever a unicode string is passed.

yajo commented on event_session/wizards/wizard_event_session.py:
```diff
@@ -0,0 +1,251 @@
+        data = {
+            "name": "{} {} - {}".format(
+                strftime('%A %d/%m/%y',
+                         strptime(values["date"], "%Y-%m-%d %H:%M:%S")),
```
All of this makes not much sense IMHO. You should just use the current context lang's settings instead of inventing all of this.

Hint: `self.env["res.lang"]._lang_get(self.env.lang)` fields: `date_format` and `time_format`.

chienandalu replied:
Well, but I really want the session name formatted like that for usability's sake

yajo commented on event_session/wizards/wizard_event_session.py:
```diff
@@ -0,0 +1,251 @@
+                strftime('%A %d/%m/%y',
+                         strptime(values["date"], "%Y-%m-%d %H:%M:%S")),
+                "%02d:%02d" % divmod(values["start_time"]*60, 60),
+                "%02d:%02d" % divmod(values["end_time"]*60, 60),
```
I guess this works, although it is a little bit hacky to read. Once you have `date` and `lang` from above, you can do this to get it perfect:

`strftime(lang.time_format, date + datetime.timedelta(seconds=values["start_time"])`

chienandalu replied:
Yes, a rework could be made with all the hours and dates management.

yajo commented on event_session/wizards/wizard_event_session.py:
```diff
@@ -0,0 +1,251 @@
+            "date_end": '%s : %s' % (
+                strftime('%Y-%m-%d', strptime(
+                    values["date"], "%Y-%m-%d %H:%M:%S")),
+                "%02d:%02d" % divmod(values["end_time"]*60, 60)),
```
Possibly some of the above clues will make this better too.

yajo commented on event_session/wizards/wizard_event_session_view.xml:
```diff
@@ -0,0 +1,74 @@
+                        type="object"
+                        string="Generate sessions"
+                        class="oe_highlight"/>
+                or
```
Drop this line.

yajo commented on event_session/models/event_session.py:
```diff
@@ -0,0 +1,195 @@
+
+
+class EventSession(models.Model):
+    _name = 'event.session'
```
What's the point on creating this model instead of using upstream `event.track`?

chienandalu replied:
We wanted something like pseudo-events. Another approach was to group events with something like a super-event, but was discarded.

pedrobaeza replied:
Tracks have another goal: they are part of an event. You can even have an event with several tracks and several sessions.

yajo replied:
Hmm I get it. It's like when you have a training of 5 tracks (monday to friday) but you repeat the training 3 weeks, right?


Interesting, although IMHO you should document more explicitly this addon's purpose, or further reviewers will wonder about this...

pedrobaeza replied:
Yeah, exactly about both points.

yajo reviewed (changes requested):
Date programming is a bit weird always, but I feel you are reinventing some wheels here. Check comments.

pedrobaeza commented on event_session/README.rst:
```diff
@@ -0,0 +1,45 @@
+   :alt: License: AGPL-3
+
+
+Create and assign sessions to events
```
above line missing

pedrobaeza replied:
The line above is still missing

pedrobaeza committed:
[IMP] event_session: Several improvements:

* Rename 'date' to 'date_begin' for being homogeneous with event fields
* Split tests
* Remove 'start_time' and 'end_time' as no needed
* Convert hours to UTC for avoiding timezone issues

sergio-teruel committed:
[IMP] event_session: Better info in registration reports

rafaelbn commented:
Travis still fails cc @sergio-teruel @chienandalu @pedrobaeza

chienandalu committed:
[IMP][FIX] event_session: several improvements

- Fix tz in calculated name
- Fix tests
- Add event sessions count to reports
- Fixes computing update of stored value
- Fixes views and reports
- Fixes generator not creating last day sessions
- Totalizations in session tree view
- Security config
- Adds es_ES translations
- Fixes seats limits on sessions

pedrobaeza commented on event_session/models/event.py:
```diff
@@ -0,0 +1,82 @@
+    @api.multi
+    @api.constrains('seats_max', 'seats_available')
+    def _check_seats_limit(self):
+        if self.sessions_count < 1:
```
This needs a `for` loop

pedrobaeza replied:
Use `if not event.session_ids:` instead (being `event` the variable for each record.

pedrobaeza commented on event_session/models/event.py:
```diff
@@ -0,0 +1,82 @@
+    def create(self, vals):
+        event = super(EventEvent, self).create(vals)
+        if not event.session_ids:
+            event.session_ids = [(0, 0, {
```
This shouldn't be needed

pedrobaeza commented on event_session/models/event.py:
```diff
@@ -0,0 +1,82 @@
+    @api.multi
+    @api.constrains('event_id', 'session_id', 'state')
+    def _check_seats_limit(self):
+        for registration in self:
```
Save one `if` putting `for registration in self.filtered('session_id'):`

pedrobaeza commented on event_session/models/event.py:
```diff
@@ -0,0 +1,82 @@
+    def confirm_registration(self):
+        for reg in self:
+            if not reg.event_id.session_ids:
+                super(EventRegistration, self).confirm_registration()
```
You are passing `self` instead of `reg` to super. Take much care with this!

pedrobaeza committed:
[FIX] event_session: Use new SQL view

pedrobaeza commented:
I have added a commit for using the new SQL view and that should fix also the need for at least one session.

chienandalu committed:
[FIX][IMP] event_session: Several improvements:

- Fix report query
- Adds progress bar to sessions tree view
- Improves tests
- Fixes seats limit check

chienandalu commented:
@Yajo @pedrobaeza @sergio-teruel This is ready for review

yajo commented on event_session/README.rst:
```diff
@@ -14,13 +14,17 @@ Usage
 * Go to Events > Sessions and create some sessions associated with an event.
-* Go to an event and use the sessions wizard to create all your event sessions according to a given schedule.
+* Go to an event and use the sessions wizard to create all your event sessions
+according to a given schedule.
```
Prefix with 2 spaces or formatting will break

yajo commented on event_session/models/event.py:
```diff
@@ -50,25 +45,25 @@ class EventRegistration(models.Model):
-                    (1 if self.state == 'draft' else 0)):
+                    registration.session_id.seats_max and
+                    registration.session_id.seats_available <
+                    (1 if registration.state == 'draft' else 0)):
```
Are you checking if we have less than zero available seats? Like -1? 🤔

chienandalu replied:
I can't recall why this was implemented like this. I've rewritten it .

yajo commented on event_session/tests/test_session.py:
```diff
@@ -90,11 +91,13 @@ def test_open_registrations(self):
-        self.assertEqual(len(self.session.event_mail_ids), 1)
-        self.session._session_mails_from_template(self.event.id, self.template)
-        self.assertEqual(len(self.session.event_mail_ids), 1)
+    # TODO: Revisar test
```
🗑

chienandalu replied:
I got to fix it

pedrobaeza commented on event_session/models/event_mail.py:
```diff
@@ -0,0 +1,86 @@
+            if (event_mail_reg.registration_id and
+                    event_mail_reg.registration_id.session_id):
+                date_open = event_mail_reg.registration_id.date_open
+                date_open_datetime = date_open and datetime.strptime(
```
Use `fields.Datetime.from_string`

pedrobaeza commented on event_session/__manifest__.py:
```diff
@@ -0,0 +1,24 @@
+    'website': 'https://odoo-community.org/',
+    'category': 'Marketing',
+    'summary': 'Sessions in events',
+    'depends': ['event', 'event_mail'],
```
I would like to avoid `event_mail` dependency here.

pedrobaeza commented on event_session/models/event_session.py:
```diff
@@ -0,0 +1,275 @@
+        seats by session.
+        """
+        # initialize fields to 0
+        for session in self:
```
This is not needed, as by default all fields are reset to their default values (0 in this case).

pedrobaeza commented on event_session/models/event_session.py:
```diff
@@ -0,0 +1,275 @@
+                'open': 'seats_reserved',
+                'done': 'seats_used',
+            }
+            query = """
```
Can you use `read_group` instead of this ugly SQL?

pedrobaeza commented on event_session/models/event_session.py:
```diff
@@ -0,0 +1,275 @@
+                    fields.Datetime.context_timestamp(self_in_tz, date_begin),
+                )
+            else:
+                session.date_begin_located = False
```
No need to assign default value, so instead just do `for session in self.filtered('date_begin'):`

pedrobaeza commented on event_session/models/event_session.py:
```diff
@@ -0,0 +1,275 @@
+                    fields.Datetime.context_timestamp(self_in_tz, date_end),
+                )
+            else:
+                session.date_end_located = False
```
The same

pedrobaeza commented on event_session/models/event_session.py:
```diff
@@ -0,0 +1,275 @@
+
+    @api.onchange('event_id')
+    def onchange_event_id(self):
+        self.seats_min = self.event_id.seats_min
```
Use `self.update` for making only one write operation (in case this is called outside an UI onchange itself).

pedrobaeza reviewed (changes requested):
And finally make a proper squash of all the commits.

rafaelbn commented:
Please @chienandalu review comment. This module is in production for a long time we should be able to merge. I'm testing it

rafaelbn reviewed (approved):
Functionally tested 👍 . Please @sergio-teruel last review here. Thanks

rafaelbn commented:
Hi @chienandalu please could `And finally make a proper squash of all the commits.` to make @pedrobaeza happy and merge this? 😄 👍

chienandalu commented:
There's quite a mess o commits from @pedrobaeza @sergio-teruel and myself :sweat_smile: Should I squash them all together or do I try to keep the closer to a history?

pedrobaeza commented:
Group them by author/logic

chienandalu commented:
Yes, the problem is that they are very mixed up in both ways. I'll try my best :)

chienandalu commented:
@pedrobaeza Now rebased (sorry for the delay)
